res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1136 @@
+        constructor(deckNumbers) {
+            super();
+            if (deckNumbers !== undefined) {
+                if (Array.isArray(deckNumbers)) {
+                    // These must be unique to each instance,
+                    // so they cannot be in the prototype.
+                    this.currentDeck = "[Channel" + deckNumbers[0] + "]";
+                    this.deckNumbers = deckNumbers;
+                } else if (typeof deckNumbers === "number" &&
+                        Math.floor(deckNumbers) === deckNumbers &&
+                        isFinite(deckNumbers)) {
+                    this.currentDeck = "[Channel" + deckNumbers + "]";
+                    this.deckNumbers = [deckNumbers];
+                }
+            } else {
+                throw "new Deck() called without specifying any deck numbers";
+            }
+        }
+        setCurrentDeck(newGroup) {
+            this.currentDeck = newGroup;
+            this.reconnectComponents(function(component) {
+                if (component.group === undefined
+                      || component.group.search(script.channelRegEx) !== -1) {
+                    component.group = newGroup;
+                } else if (component.group.search(script.eqRegEx) !== -1) {
+                    component.group = "[EqualizerRack1_" + newGroup + "_Effect1]";
+                } else if (component.group.search(script.quickEffectRegEx) !== -1) {
+                    component.group = "[QuickEffectRack1_" + newGroup + "]";
+                }
+                // Do not alter the Component's group if it does not match any of those RegExs.
+
+                if (component instanceof EffectAssignmentButton) {
+                    // The ControlObjects for assigning decks to effect units
+                    // indicate the effect unit with the group and the deck with the key,
+                    // so change the key here instead of the group.
+                    component.inKey = "group_" + newGroup + "_enable";
+                    component.outKey = "group_" + newGroup + "_enable";
+                }
+            });
+        }
+        toggle() {
+            // cycle through deckNumbers array
+            var index = this.deckNumbers.indexOf(parseInt(
+                script.channelRegEx.exec(this.currentDeck)[1]
+            ));
+            if (index === (this.deckNumbers.length - 1)) {
+                index = 0;
+            } else {
+                index += 1;
+            }
+            this.setCurrentDeck("[Channel" + this.deckNumbers[index] + "]");
+        }
+    };
+
+    var EffectUnit = class extends ComponentContainer {
+        constructor(unitNumbers, allowFocusWhenParametersHidden, colors) {
+            super();
+            var eu = this;
+            this.focusChooseModeActive = false;
+
+            // This is only connected if allowFocusWhenParametersHidden is false.
+            this.onShowParametersChange = function(value) {
```

Comment:
Shouldn't this be a method instead of function expression property? (also applies to the rest of the code)